### PR TITLE
Attribute-based approach for opting-out of sort input types

### LINKED ISSUE #4579: InputObject GeometryOverlaySortInput has no fields declared
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Graphql with Hotchocolate does not start because GeometryOverlaySortInput of NetTopologySuite package has no fields declared.

I upgraded the Hotchocolate Graphql packages and EntityFramework Core packages, and now the application starts but graphql is not working. I tried downgrading the versions one by one, but it seems that reversing to ef core packages version 5.0.13 and lower this error doesn't happen. Maybe you can show me how can I contribute to fix this issue and I will do a PR.

Thanks a lot

### Steps to reproduce

1. Upgrade EntityFrameworkCore dependencies to 6.0.0 or 6.0.1
2. Include EntityFrameworkCore  NetTopologySuite library
3. Start GraphQL host instance

### Relevant log output

```shell
1. InputObject `GeometryOverlaySortInput` has no fields declared. (HotChocolate.Data.Sorting.SortInputType<NetTopologySuite.Geometries.GeometryOverlay>)
2. InputObject `CoordinateEqualityComparerSortInput` has no fields declared. (HotChocolate.Data.Sorting.SortInputType<NetTopologySuite.Geometries.CoordinateEqualityComparer>)

   at HotChocolate.Configuration.TypeInitializer.EnsureNoErrors()
   at HotChocolate.Configuration.TypeInitializer.CompleteTypes()
   at HotChocolate.Configuration.TypeInitializer.Initialize()
   at HotChocolate.SchemaBuilder.Setup.InitializeTypes(SchemaBuilder builder, IDescriptorContext context, IReadOnlyList`1 types, LazySchema lazySchema)
   at HotChocolate.SchemaBuilder.Setup.Create(SchemaBuilder builder, LazySchema lazySchema, IDescriptorContext context)
   at HotChocolate.SchemaBuilder.Create(IDescriptorContext context)
   at HotChocolate.SchemaBuilder.HotChocolate.ISchemaBuilder.Create(IDescriptorContext context)
   at HotChocolate.Execution.RequestExecutorResolver.CreateSchemaAsync(NameString schemaName, RequestExecutorSetup options, RequestExecutorOptions executorOptions, IServiceProvider serviceProvider, TypeModuleChangeMonitor typeModuleChangeMonitor, CancellationToken cancellationToken)
   at HotChocolate.Execution.RequestExecutorResolver.CreateSchemaServicesAsync(NameString schemaName, RequestExecutorSetup options, CancellationToken cancellationToken)
   at HotChocolate.Execution.RequestExecutorResolver.GetRequestExecutorNoLockAsync(NameString schemaName, CancellationToken cancellationToken)
   at HotChocolate.Execution.RequestExecutorResolver.GetRequestExecutorAsync(NameString schemaName, CancellationToken cancellationToken)
   at HotChocolate.Execution.RequestExecutorProxy.GetRequestExecutorAsync(CancellationToken cancellationToken)
   at HotChocolate.AspNetCore.HttpPostMiddleware.HandleRequestAsync(HttpContext context, AllowedContentType contentType)
   at HotChocolate.AspNetCore.HttpPostMiddleware.InvokeAsync(HttpContext context)
   at HotChocolate.AspNetCore.WebSocketSubscriptionMiddleware.InvokeAsync(HttpContext context)
   at Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions.<>c__DisplayClass13_0.<<UseCancellation>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```
```


### Additional Context?

EF Core version: 6.0.1
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 6.0
Operating system: Windows
IDE: JetBrains Rider

### Product

Hot Chocolate

### Version

12.4.1

### PR DESCRIPTION
**Pull request by willwolfram18**
Added an attribute-based approach for removing fields from `HotChocolate.Data.Sorting` scenarios.

- Introduced a new `[GraphQLIgnoreSort]` attribute, which is looked for by `SortInputTypeDescriptor<T>`
- Implemented new test for `[GraphQLIgnoreSort]` attribute
- Updated v12 and v13 docs to describe the attribute

Closes #4579
- I realize the above issue is already closed. This implementation provides an alternative approach to address [this comment](https://github.com/ChilliCream/hotchocolate/issues/4579#issuecomment-1351946124) without needing to define a custom `SortInputTypeDescriptor<T>`

### CODE REVIEW

**Comment by michaelstaib** [comment]
The issue I have with this is that we need for a ton of other things ignore attributes which I think is not really feasible. We have not tackled this because so far nothing felt right.

A single entity here could need ignores for Inputs, Interface, Objects, Storting, Filtering and other things ...

**Comment by willwolfram18** [comment]
> The issue I have with this is that we need for a ton of other things ignore attributes which I think is not really feasible. We have not tackled this because so far nothing felt right.
> 
> A single entity here could need ignores for Inputs, Interface, Objects, Storting, Filtering and other things ...

Yeah, I can understand that. I will say that this does feel less cumbersome than the approach that's currently available in the library though, at least in regards to sort and filter types. The current state requires:

* Define a custom `SortInputType<T>` and/or `FilterInputType<T>`
* Override the `Configure(ISortInputTypeDescriptor<T>)` or `Configure(IFilterInputTypeDescriptor<T>)` method
* Configure the implicit field bindings (done be default so not explicitly required to code this)
* Call the `descriptor.Ignore(x => ...)` method to remove the offending attribute(s) from the input type
* Explicitly register the new type with the `AddType<T>()` method on the schema builder

So if someone wants to remove a specific field from filtering and sorting, they need to define multiple types and register them with the schema. And if you want to ignore a particular **type altogether**, it gives even more complicated since you need to create custom types for each separate type that has a field using that type.

The attribute-based approach can accomplish the same capability, at least for sort and filter input types, with a lot less effort, which is why I figured that might be a nice addition to the library for code-first implementations 👍 But with that said, I'm not as familiar with some of the other use cases and scenarios that such an `Ignore()` approach might need for other types and how well this idea would scale ☹️ 

I'm happy to leave this PR open, convert it to a draft, or close it altogether if the HC team feels more thought and discussion is needed on this idea.

**Comment by michaelstaib** [comment]
I will have a look at it over the next days and see if we can find a way to allow that which covers all cases.

File: website/src/docs/hotchocolate/v12/fetching-data/sorting.md [hunk 1]
```diff
@@ -155,7 +155,7 @@ Under the hood, sorting is based on top of normal Hot Chocolate input types. You
 `ISortInputTypeDescriptor<T>` supports most of the methods of `IInputTypeDescriptor<T>`. By default, operations are generated for all fields of the type.
 Members that are collections are skipped because you cannot order based on lists.
 If you do want to specify the sorting types by yourself, you can change this behavior with `BindFields`, `BindFieldsExplicitly`, or `BindFieldsImplicitly`.
-When fields are bound implicitly, meaning sorting is added for all valid properties, you may want to hide a few fields. You can do this with `Ignore(x => Bar)`.
+When fields are bound implicitly, meaning sorting is added for all valid properties, you may want to hide a few fields. You can do this with `Ignore(x => Bar)`. You an also use the `[GraphQLIgnoreSort]` attribute on a particular property to explicitly exclude that field, or place it on a class definition to remove all fields from sort inputs that use that type.
```
Comment (wayne-schwebel-cs) [inline website/src/docs/hotchocolate/v12/fetching-data/sorting.md:158]:
```suggestion
When fields are bound implicitly, meaning sorting is added for all valid properties, you may want to hide a few fields. You can do this with `Ignore(x => Bar)`. You can also use the `[GraphQLIgnoreSort]` attribute on a particular property to explicitly exclude that field, or place it on a class definition to remove all fields from sort inputs that use that type.
```

File: website/src/docs/hotchocolate/v13/fetching-data/sorting.md [hunk 2]
```diff
@@ -155,7 +155,7 @@ Under the hood, sorting is based on top of normal Hot Chocolate input types. You
 `ISortInputTypeDescriptor<T>` supports most of the methods of `IInputTypeDescriptor<T>`. By default, operations are generated for all fields of the type.
 Members that are collections are skipped because you cannot order based on lists.
 If you do want to specify the sorting types by yourself, you can change this behavior with `BindFields`, `BindFieldsExplicitly`, or `BindFieldsImplicitly`.
-When fields are bound implicitly, meaning sorting is added for all valid properties, you may want to hide a few fields. You can do this with `Ignore(x => Bar)`.
+When fields are bound implicitly, meaning sorting is added for all valid properties, you may want to hide a few fields. You can do this with `Ignore(x => Bar)`. You an also use the `[GraphQLIgnoreSort]` attribute on a particular property to explicitly exclude that field, or place it on a class definition to remove all fields from sort inputs that use that type.
```
Comment (wayne-schwebel-cs) [inline website/src/docs/hotchocolate/v13/fetching-data/sorting.md:158]:
```suggestion
When fields are bound implicitly, meaning sorting is added for all valid properties, you may want to hide a few fields. You can do this with `Ignore(x => Bar)`. You can also use the `[GraphQLIgnoreSort]` attribute on a particular property to explicitly exclude that field, or place it on a class definition to remove all fields from sort inputs that use that type.
```

**Comment by michaelstaib** [comment]
Closing this PR as we do not find the solution to be optimal.

**Pull request closed without merging**